service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateAzureIpStep.java

```diff
@@ -0,0 +1,118 @@
+package bio.terra.workspace.service.resource.controlled.flight.create;
+
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.CREATION_PARAMETERS;
+
+import bio.terra.cloudres.azure.resourcemanager.common.Defaults;
+import bio.terra.cloudres.azure.resourcemanager.compute.data.CreatePublicIpRequestData;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.app.configuration.external.AzureConfiguration;
+import bio.terra.workspace.generated.model.ApiAzureIpCreationParameters;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.ControlledAzureIpResource;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.resourcemanager.compute.ComputeManager;
+import com.azure.resourcemanager.network.models.IpAllocationMethod;
+import com.azure.resourcemanager.network.models.PublicIpAddress;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateAzureIpStep implements Step {
+
+  private static final Logger logger = LoggerFactory.getLogger(CreateGcsBucketStep.class);
+  private final AzureConfiguration azureConfig;
+  private final CrlService crlService;
+  private final ControlledAzureIpResource resource;
+  private final WorkspaceService workspaceService;
+  private final AzureCloudContext azureCloudContext;
+
+  public CreateAzureIpStep(
+      AzureConfiguration azureConfig,
+      AzureCloudContext azureCloudContext,
+      CrlService crlService,
+      ControlledAzureIpResource resource,
+      WorkspaceService workspaceService) {
+    this.azureConfig = azureConfig;
+    this.azureCloudContext = azureCloudContext;
+    this.crlService = crlService;
+    this.resource = resource;
+    this.workspaceService = workspaceService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    FlightMap inputMap = context.getInputParameters();
+    ApiAzureIpCreationParameters creationParameters =
+        inputMap.get(CREATION_PARAMETERS, ApiAzureIpCreationParameters.class);
+
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+
+    // Don't try to create it if it already exists. At this point the assumption is
```

Comment:
I think that is a faulty assumption. You cannot tell the difference between a previously existing IP - where you should not be creating it - and a rerun of a valid create attempt.

The consequence is that the undo will not behave properly. It will leak IPs if the step is rerun and then a later step errors.

This really needs to be split into two steps. One step to determine whether the IP exists and throws if it does.
Then this step always knows for sure that if the IP exists, it is a rerun.